### PR TITLE
Document or-patterns

### PR DESCRIPTION
**Pull request by mark-i-m**
cc https://github.com/rust-lang/rust/pull/79278 https://github.com/rust-lang/rust/issues/54883 @joshtriplett 

@petrochenkov This also allows the leading vert change according to the T-lang FCP in https://github.com/rust-lang/rust/issues/81415

### CODE REVIEW

**Comment by ehuss** [comment]
@mark-i-m Now that the stabilization PR has merged, is this up-to-date?  It has been a bit difficult to keep up with all the changes.

File: src/patterns.md [hunk 1]
```diff
@@ -2,8 +2,13 @@
 
 > **<sup>Syntax</sup>**\
 > _Pattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `|`<sup>?</sup> _PatternNoTopAlt_ `|` _PatternNoTopAlt_
+> &nbsp;&nbsp; | _PatternNoTopAlt_
+>
+> _PatternNoTopAlt_ :\
 > &nbsp;&nbsp; &nbsp;&nbsp; _PatternWithoutRange_\
 > &nbsp;&nbsp; | [_RangePattern_]
+> &nbsp;&nbsp; | `(` _Pattern_ `)`
```
Comment (ehuss) [inline src/patterns.md:11]: Need to be careful with trailing backslashes which indicate a hard newline.

```suggestion
> &nbsp;&nbsp; | [_RangePattern_]\
> &nbsp;&nbsp; | `(` _Pattern_ `)`
```

File: src/patterns.md [hunk 2]
```diff
@@ -2,8 +2,13 @@
 
 > **<sup>Syntax</sup>**\
 > _Pattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `|`<sup>?</sup> _PatternNoTopAlt_ `|` _PatternNoTopAlt_
+> &nbsp;&nbsp; | _PatternNoTopAlt_
```
Comment (ehuss) [inline src/patterns.md:6]: I'm not sure I see how this grammar allows more than 2 repetitions.  I would maybe expect something like this:

```suggestion
> &nbsp;&nbsp; &nbsp;&nbsp; `|`<sup>?</sup> _PatternNoTopAlt_ ( `|` _PatternNoTopAlt_ )<sup>\*
```

File: src/expressions/closure-expr.md [hunk 3]
```diff
@@ -10,7 +10,7 @@
 > &nbsp;&nbsp; _ClosureParam_ (`,` _ClosureParam_)<sup>\*</sup> `,`<sup>?</sup>
 >
 > _ClosureParam_ :\
-> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> [_Pattern_]&nbsp;( `:` [_Type_] )<sup>?</sup>
+> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> [_PatternNoTopAlt_]&nbsp;( `:` [_Type_] )<sup>?</sup>
```
Comment (ehuss) [inline src/expressions/closure-expr.md:13]: Need to replace the reference link at the bottom of the page for `[_Pattern_]` to be `[_PatternNoTopAlt_]`.

File: src/items/functions.md [hunk 4]
```diff
@@ -32,7 +32,7 @@
 > )
 >
 > _FunctionParamPattern_ :\
-> &nbsp;&nbsp; [_Pattern_] `:` ( [_Type_] | `...` )
+> &nbsp;&nbsp; [_PatternNoTopAlt_] `:` ( [_Type_] | `...` )
```
Comment (ehuss) [inline src/items/functions.md:35]: Same here about the link.

File: src/statements.md [hunk 5]
```diff
@@ -53,7 +53,7 @@ fn outer() {
 
 > **<sup>Syntax</sup>**\
 > _LetStatement_ :\
-> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_Pattern_]
+> &nbsp;&nbsp; [_OuterAttribute_]<sup>\*</sup> `let` [_PatternNoTopAlt_]
```
Comment (ehuss) [inline src/statements.md:56]: Same here about the link.

File: src/patterns.md [hunk 6]
```diff
@@ -756,6 +761,57 @@ Path patterns are irrefutable when they refer to structs or an enum variant when
...
 refer to refutable constants or enum variants for enums with multiple variants.
 
+## Or-patterns
+
+_Or-patterns_ are patterns that match on either of two sub-patterns (e.g. `A |
+B`). They can nest arbitrarily. Syntactically, or-patterns are allowed in any
+of the places where other patterns are allowed (represented by the _Pattern_
+production), with the exceptions of function and closure arguments (represented
+by the _PatternNoTopAlt_ production). Additionally, the macro `:pat` matchers
+only match or-patterns in the 2021+ editions.
```
Comment (ehuss) [inline src/patterns.md:771]: I would just leave out this point about macros, and instead just document in macros-by-example.md.  Also, the reference has a style of using blockquotes to call out edition differences with `> **Edition Differences**:`, so I would expect something similar in the macros page.

```suggestion
by the _PatternNoTopAlt_ production).
```

File: src/patterns.md [hunk 7]
```diff
@@ -756,6 +761,57 @@ Path patterns are irrefutable when they refer to structs or an enum variant when
 has only one variant or a constant whose type is irrefutable. They are refutable when they
 refer to refutable constants or enum variants for enums with multiple variants.
 
+## Or-patterns
+
+_Or-patterns_ are patterns that match on either of two sub-patterns (e.g. `A |
+B`). They can nest arbitrarily. Syntactically, or-patterns are allowed in any
```
Comment (ehuss) [inline src/patterns.md:767]: I'm not sure I follow what this explicitly mentions "two" sub-patterns.  Can't it be any number?

File: src/patterns.md [hunk 8]
```diff
@@ -756,6 +761,57 @@ Path patterns are irrefutable when they refer to structs or an enum variant when
...
+
+1. Given a pattern `p | q` at some depth for some arbitrary patterns `p` and `q`,
+   the pattern is considered ill-formed if:
+
+   + the type inferred for `p` does not unify with the type inferred for `q`, or
+   + the same set of bindings are not introduced in `p` and `q`, or
+   + the type of any two bindings with the same name in `p` and `q` do not unify
+     with respect to types or binding modes.
+
+   [type coercions]: https://doc.rust-lang.org/reference/type-coercions.html
```
Comment (ehuss) [inline src/patterns.md:783]: Links should always be relative, and preferably placed at the bottom of the page.

```suggestion
   [type coercions]: type-coercions.md
```

**Review by ehuss — COMMENTED** [review]
I believe `_MatchArmPatterns_` can now be replaced with `_Pattern_` correct?

The `pat` entry in macros-by-example.md needs to be updated.

Can you add some discussion of the precedence issues (particularly with `@`, though I'm uncertain what else needs to be said)?

**Comment by mark-i-m** [comment]
Sorry for the delay. I will try to get to this soon.

At a high level, yes, I think this is up to date, though skimming your comments, I think you are right about some errors. Much of this is copy/pasted from the RFC.

File: src/patterns.md (hunk 2 above)
Comment (mark-i-m) [reply src/patterns.md:6]: I think you are right. Fixed.

File: src/patterns.md (hunk 7 above)
Comment (mark-i-m) [reply src/patterns.md:767]: I had copy-pasted from the RFC without reading carefully. Fixed.

**Comment by mark-i-m** [comment]
> Can you add some discussion of the precedence issues (particularly with @, though I'm uncertain what else needs to be said)?

Sure, unfortunately, I have to run atm... will come back to this.

**Comment by mark-i-m** [comment]
Ok, @ehuss I think I have addressed everything. Please let me know what you think. Thanks!

File: src/macros-by-example.md [hunk 9]
```diff
@@ -122,7 +122,9 @@ fragment specifiers are:
   * `block`: a [_BlockExpression_]
   * `stmt`: a [_Statement_] without the trailing semicolon (except for item
     statements that require semicolons)
-  * `pat`: a [_Pattern_]
+  * `pat2015`: a [_PatternNoTopAlt_]
+  * `pat2021`: a [_Pattern_]
+  * `pat`: equivalent to either `pat2015` or `pat2021`, depending on edition (see note below).
```
Comment (ehuss) [inline src/macros-by-example.md:127]: IIUC, `path2015` and `pat2021` are still unstable, correct?  If so, can you remove them? The reference only documents the stable language.

File: src/macros-by-example.md [hunk 10]
```diff
@@ -133,6 +135,11 @@ fragment specifiers are:
   * `vis`: a possibly empty [_Visibility_] qualifier
   * `literal`: matches `-`<sup>?</sup>[_LiteralExpression_]
 
+> **Edition Differences**: In Editions 2015 and 2018, `pat` fragment-specifiers
+> do not allow top-level or-patterns (i.e. they accept [_PatternNoTopAlt_]). In
+> Edition 2021, this was changed so that top-level or-patterns are accepted by
+> `pat` (i.e. [_Pattern_]).
+
```
Comment (ehuss) [inline src/macros-by-example.md:142]: Also remove this.  2021 changes will be added when 2021 is stabilized.

**Comment by mark-i-m** [comment]
@ehuss I updated the PR to remove unstable stuff. I'll open another PR with the 2021 stuff that can be merged later.

Also, in https://github.com/rust-lang/rust/pull/83386 there is an FCP (almost over) to stabilize `pat2015`, except it was renamed to `pat_param`. I've made that change in my last PR and left `pat_param` in the grammar. Let me know if this is ok.

**Comment by ehuss** [comment]
I removed `pat_param` since it isn't clear if #83386 will merge in time.  I'll add it back whenever it does merge.

**Review by ehuss — APPROVED** [review]
Thanks!

**Pull request merged**